proxy/src/auth/backend.rs

```diff
@@ -214,14 +215,48 @@ async fn auth_quirks(
         Some(secret) => secret,
         None => api.get_role_secret(ctx, &info).await?,
     };
+    let (cached_entry, secret) = cached_secret.take_value();
+
+    let secret = match secret {
+        Some(secret) => {
+            // we have validated the endpoint exists, so let's intern it.
+            let endpoint = EndpointIdInt::from(&info.endpoint);
+
+            // only count the full hash count if password hack or websocket flow.
+            // in other words, if proxy needs to run the hashing
+            let password_weight = if unauthenticated_password.is_some() || allow_cleartext {
+                match &secret {
+                    #[cfg(any(test, feature = "testing"))]
+                    AuthSecret::Md5(_) => 1,
+                    AuthSecret::Scram(s) => s.iterations + 1,
+                }
+            } else {
+                // validating scram takes just 1 hmac_sha_256 operation.
+                1
+            };
+
+            if config
+                .rate_limiter
+                .check((endpoint, ctx.peer_addr), password_weight)
+            {
+                secret
+            } else {
+                // If we don't have an authentication secret, we mock one to
+                // prevent malicious probing (possible due to missing protocol steps).
+                // This mocked secret will never lead to successful authentication.
+                warn!("rate limiting authentication, mocking it");
```

Comment:
I guess, here should be reject + log + record to metric, right?

Reply:
Yep. On it